File: src/ios/FirebasePlugin.m
#import "FirebasePlugin.h"
#import "Firebase.h"
#import "AppDelegate.h"
#import <Cordova/CDVPlugin.h>
#import <Crashlytics/Crashlytics.h>
#import <Fabric/Fabric.h>

@implementation FirebasePlugin

- (void)admobRequestInterstitial {
    [self.commandDelegate runInBackground:^{
        self.interstitial = [[GADInterstitial alloc] initWithAdUnitID:self.interstitialId];
        self.interstitial.delegate = (id <GADInterstitialDelegate>)self;

        GADRequest *request = [GADRequest request];

        request.testDevices = self.testDevices;

        [self.interstitial loadRequest:request];
    }];
}

- (void)admobSetup:(CDVInvokedUrlCommand*)command {
    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    NSString* appId = [command.arguments objectAtIndex:0];
    NSString* interstitialId = [command.arguments objectAtIndex:1];
    NSArray* testDevices = [command.arguments objectAtIndex:2];

    if (testDevices && testDevices.count) {
        [self.testDevices addObjectsFromArray:testDevices];
    }

    [GADMobileAds configureWithApplicationID:appId];

    self.applicationId = appId;
    self.interstitialId = interstitialId;

    [self admobRequestInterstitial];

    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)admobShowInterstitial:(CDVInvokedUrlCommand*)command {
    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];

    [self.commandDelegate runInBackground:^{
        if (self.interstitial.isReady) {
            [self.interstitial presentFromRootViewController:self.viewController];
        }

        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }];
}

- (void)analyticsLogEvent:(CDVInvokedUrlCommand*)command {
    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    NSString* name = [command.arguments objectAtIndex:0];
    NSDictionary *parameters;

    @try {
        NSString *description = NSLocalizedString([command argumentAtIndex:1 withDefault:@"No Message Provided"], nil);

        parameters = @{ NSLocalizedDescriptionKey: description };
    }
    @catch (NSException *execption) {
        parameters = [command argumentAtIndex:1];
    }

    [self.commandDelegate runInBackground:^{
        [FIRAnalytics logEventWithName:name parameters:parameters];

        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }];
}

- (void)analyticsSetUserId:(CDVInvokedUrlCommand*)command {
    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    NSString* userId = [command.arguments objectAtIndex:0];

    [self.commandDelegate runInBackground:^{
        [FIRAnalytics setUserID:userId];

        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }];
}

- (void)analyticsSetScreenName:(CDVInvokedUrlCommand*)command {
    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    NSString* name = [command.arguments objectAtIndex:0];

    [self.commandDelegate runInBackground:^{
        [FIRAnalytics setScreenName:name screenClass:NULL];

        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }];
}

- (void)analyticsSetUserProperty:(CDVInvokedUrlCommand*)command {
    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    NSString* name = [command.arguments objectAtIndex:0];
    NSString* value = [command.arguments objectAtIndex:1];

    [self.commandDelegate runInBackground:^{
        [FIRAnalytics setUserPropertyString:name forName:value];

        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }];
}

- (void)crashlyticsTest:(CDVInvokedUrlCommand*)command {
    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];

    [self.commandDelegate runInBackground:^{
        [[Crashlytics sharedInstance] crash];

        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }];
}

- (void)interstitialDidDismissScreen:(GADInterstitial *)interstitial {
    NSLog(@"POOPING");
    [self admobRequestInterstitial];
}


- (void)remoteConfigSetup:(CDVInvokedUrlCommand*)command {
    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];

    [self.commandDelegate runInBackground:^{
        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }];
}

@end

File: src/ios/FirebasePlugin.swift
import Crashlytics
import Firebase
import GoogleMobileAds

var testDevices: [String] = []

@objc(FirebasePlugin)
class FirebasePlugin : CDVPlugin {

    var applicationId: String = "ca-app-pub-3940256099942544~1458002511"
    var interstitial: GADInterstitial!
    var interstitialUnitId: String = "ca-app-pub-3940256099942544/4411468910"


    @objc(pluginInitialize)
    override func pluginInitialize() {
        if (FirebaseApp.app() == nil) {
            FirebaseApp.configure()
        }
    }

    @objc(admobAddTestDevice:)
    func admobAddTestDevice(command: CDVInvokedUrlCommand) {
        let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK)
        let testDeviceId = command.arguments[0] as? String ?? ""

        if !testDeviceId.isEmpty {
            testDevices.append(testDeviceId)
        }

        self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
    }

    @objc(admobGetTestDevices:)
    func admobGetTestDevices(command: CDVInvokedUrlCommand) {
        self.commandDelegate.send(
            CDVPluginResult(status: CDVCommandStatus_OK, messageAs: testDevices),
            callbackId: command.callbackId
        )
    }

    @objc(admobRequestInterstitial)
    func admobRequestInterstitial() {
        DispatchQueue.global(qos: .userInitiated).async {
            self.interstitial = GADInterstitial(adUnitID: self.interstitialUnitId)

            let request = GADRequest()

            request.testDevices = testDevices

            self.interstitial.load(request)
        }
    }

    @objc(admobSetAdmobAppId:)
    func admobSetAdmobAppId(command: CDVInvokedUrlCommand) {
        DispatchQueue.global(qos: .userInitiated).async {
            let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK)

            self.applicationId = command.arguments[0] as? String ?? "ca-app-pub-3940256099942544~1458002511"

            DispatchQueue.main.async {
                GADMobileAds.configure(withApplicationID: self.applicationId)

                self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
            }
        }
    }

    @objc(admobSetInterstitialId:)
    func admobSetInterstitialId(command: CDVInvokedUrlCommand) {
        DispatchQueue.global(qos: .userInitiated).async {
            let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK)

            self.interstitialUnitId = command.arguments[0] as? String ?? "ca-app-pub-3940256099942544/4411468910"

            self.admobRequestInterstitial()

            self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
        }
    }

    @objc(admobShowInterstitial:)
    func admobShowInterstitial(command: CDVInvokedUrlCommand) {
        DispatchQueue.global(qos: .userInitiated).async {
            var pluginResult = CDVPluginResult(status: CDVCommandStatus_OK)

            DispatchQueue.main.async {
                if self.interstitial != nil && self.interstitial.isReady {
                    self.interstitial.present(fromRootViewController: self.viewController)
                } else {
                    pluginResult = CDVPluginResult(status: CDVCommandStatus_ERROR)
                }

                self.admobRequestInterstitial()

                self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
            }
        }
    }

    @objc(analyticsLogEvent:)
    func analyticsLogEvent(command: CDVInvokedUrlCommand) {
        DispatchQueue.global(qos: .userInitiated).async {
            let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK)
            let eventName = command.arguments[0] as? String ?? ""
            let eventParameters = command.arguments[1] as? [String: Any] ?? nil

            Analytics.logEvent(eventName, parameters: eventParameters)

            self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
        }
    }

    @objc(analyticsSetUserId:)
    func analyticsSetUserId(command: CDVInvokedUrlCommand) {
        DispatchQueue.global(qos: .userInitiated).async {
            let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK)
            let userId = command.arguments[0] as? String ?? ""

            Analytics.setUserID(userId)

            self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
        }
    }

    @objc(crashlyticsTest:)
    func crashlyticsTest(command: CDVInvokedUrlCommand) {
        DispatchQueue.global(qos: .userInitiated).async {
            Crashlytics.sharedInstance().crash()
        }
    }

}
